core/contracts/oracle/implementation/Voting.sol

```diff
@@ -363,8 +363,8 @@ contract Voting is Testable, Ownable, OracleInterface, VotingInterface, Encrypte
         // 0 hashes are disallowed in the commit phase, so they indicate a different error.
         // Cannot reveal an uncommitted or previously revealed hash
         require(voteSubmission.commit != bytes32(0), "Invalid hash reveal");
-        // Committed hash doesn't match revealed price and salt
-        require(keccak256(abi.encode(price, salt)) == voteSubmission.commit, "Invalid commit hash & salt");
+        // Committed hash doesn't match revealed voter address, price and salt
+        require(keccak256(abi.encodePacked(price, salt, msg.sender)) == voteSubmission.commit, "Invalid commit hash, address & salt");
```

Comment:
`abi.encode(...., address)` does not return the same hash that `web3.utils.soliditySha3(....,address)` does, but `abi.encodePacked` does. I'm not 100% sure why this works, but I believe that if an address is input into `abi.encode()`, its format needs to be manipulated a bit, including getting rid of the leading `0x...`

https://ethereum.stackexchange.com/questions/29139/how-does-solidity-tightly-packed-arguments-work-in-sha256/46161

Reply:
My understanding is that `encode` and `encodePacked` differ for types that are less than 32 bytes or dynamically sized. Since an address is less than 32 bytes, I'm guessing it's creating the problem. See [here](https://solidity.readthedocs.io/en/v0.6.6/abi-spec.html#non-standard-packed-mode) for the solidity docs around this. In the [docs](https://web3js.readthedocs.io/en/v1.2.0/web3-utils.html#soliditysha3) for `soliditySha3`, they say that they "tightly pack" the arguments before hashing. So I think this is actually more correct than it was before.

Now that we're tossing more types in here, we may need to be more careful about how we compute `soliditySha3` offchain. web3 attempts to autodetect types, but it's not perfect. Since solidity knows the types, if web3 autodetects incorrectly, it could lead to hash differences due to differences in encoding. To solve this problem, web3 offers a way to provide the types of the arguments passed to `soliditySha3`. We don't need to do this in this PR, but it's something we may consider doing in the future to make sure these hashes are always computed in exactly the same way.

Reply:
Right. I'll raise an issue for this so we can do in a subsequent PR.

Reply:
Just to add one more note on this.

I want to explain my concerns around js autodecting types. If you look at the docs, it autodetects bytes, but not addresses. My guess is that addresses just get encoded the same way as bytes, which probably works fine most of the time. However, ~1 out of every 128 addresses has a 0 for the last byte (or first byte), assuming they're evenly distributed. If the string representation of that address drops the trailing 0, then it might be interpreted as a `bytes19` by js. However, solidity would still encode that 0 because it knows an address is 20 bytes. This would lead to a different hash in solidity and js. This could similarly cause problems when hashing the `bytes32` representation of our identifiers since there will be padded 0s on the end, which may not show up in the string or may be thrown out by web3. I haven't tested any of this, but these sorts of errors are plausible to the point where I think it might make sense to specify the types on the client side to avoid corrupted hashes.

Reply:
https://github.com/UMAprotocol/protocol/issues/1219

Reply:
Perfect, thanks @nicholaspai. To keep this succinct, don't worry about it in this PR unless it affects the tests.

Reply:
> Just to add one more note on this.
> 
> I want to explain my concerns around js autodecting types. If you look at the docs, it autodetects bytes, but not addresses. My guess is that addresses just get encoded the same way as bytes, which probably works fine most of the time. However, ~1 out of every 128 addresses has a 0 for the last byte (or first byte), assuming they're evenly distributed. If the string representation of that address drops the trailing 0, then it might be interpreted as a `bytes19` by js. However, solidity would still encode that 0 because it knows an address is 20 bytes. This would lead to a different hash in solidity and js. This could similarly cause problems when hashing the `bytes32` representation of our identifiers since there will be padded 0s on the end, which may not show up in the string or may be thrown out by web3. I haven't tested any of this, but these sorts of errors are plausible to the point where I think it might make sense to specify the types on the client side to avoid corrupted hashes.

Agreed with this. Will add comments to issue.

Reply:
what about type casing the input concatenation of `price + salt + voter` to a string and then it will always be the same, irrespective of type or how js interprets the leading (or trailing) 0 in the address?

Reply:
@chrismaree that would work but it would require changing the contract interface. Its much easier IMO to just make the `client` code specify the types to hash:

old code:
```
const newHash = web3.utils.soliditySha3(
     newPrice,
     newSalt,
     account1,
     time,
     currentRoundId,
     identifier
);
```
new code:
```
    const newHash = web3.utils.soliditySha3(
      { t: "int", v: newPrice },
      { t: "int", v: newSalt },
      { t: "address", v: account1 },
      { t: "uint", v: time },
      { t: "uint", v: currentRoundId },
      { t: "bytes32", v: identifier }
    );
```

Reply:
@chrismaree while unlikely, one minor issue with the concatenation approach is that it *technically* can create hash collisions between different inputs:
```
price1 = 15
salt1 = 100
price1 + salt1 = 15100
price2 = 1
salt2 = 5100
price2 + salt2 = 15100
```

I think fixed length inputs are always safest for this reason. We should have the inputs to the hash always include all their bits, even if there are leading or trailing zeros.

Reply:
great point @mrice32